src/main/scala/mesosphere/mesos/ResourceUtil.scala

```diff
@@ -1,121 +1,150 @@
+        diskSource.path).mkString(":")
+    }
+  }
 
   private[this] val log = LoggerFactory.getLogger(getClass)
 
   /**
     * The resources in launched tasks, should
     * be consumed from resources in the offer with the same [[ResourceMatchKey]].
     */
-  private[this] case class ResourceMatchKey(
+  private[this] case class ResourceConsumeKey(
     role: String, name: String,
     reservation: Option[ReservationInfo], disk: Option[DiskInfo])
-  private[this] object ResourceMatchKey {
-    def apply(resource: MesosProtos.Resource): ResourceMatchKey = {
+
+  private[this] object ResourceConsumeKey {
+    def apply(resource: MesosProtos.Resource): ResourceConsumeKey = {
       val reservation = if (resource.hasReservation) Some(resource.getReservation) else None
       val disk = if (resource.hasDisk) Some(resource.getDisk) else None
-      ResourceMatchKey(resource.getRole, resource.getName, reservation, disk)
+      ResourceConsumeKey(resource.getRole, resource.getName, reservation, disk)
     }
   }
 
   /**
-    * Deduct usedResource from resource. If nothing is left, None is returned.
+    * Decrements the scalar resource by amount
+    *
     */
-  def consumeResource(
-    resource: MesosProtos.Resource,
-    usedResource: MesosProtos.Resource): Option[MesosProtos.Resource] = {
-    require(resource.getType == usedResource.getType)
-
-    def consumeScalarResource: Option[MesosProtos.Resource] = {
-      val leftOver: Double = resource.getScalar.getValue - usedResource.getScalar.getValue
-      if (leftOver <= 0) {
-        None
-      } else {
-        Some(resource
-          .toBuilder
-          .setScalar(
-            MesosProtos.Value.Scalar
-              .newBuilder().setValue(leftOver))
-          .build())
-      }
-    }
-
-    def deductRange(
-      baseRange: MesosProtos.Value.Range,
-      usedRange: MesosProtos.Value.Range): Seq[MesosProtos.Value.Range] = {
-      if (baseRange.getEnd < usedRange.getBegin || baseRange.getBegin > usedRange.getEnd) {
-        // baseRange completely before or after usedRange
-        Seq(baseRange)
-      } else {
-        val rangeBefore: Option[MesosProtos.Value.Range] = if (baseRange.getBegin < usedRange.getBegin)
-          Some(baseRange.toBuilder.setEnd(usedRange.getBegin - 1).build())
-        else
-          None
-
-        val rangeAfter: Option[MesosProtos.Value.Range] = if (baseRange.getEnd > usedRange.getEnd)
-          Some(baseRange.toBuilder.setBegin(usedRange.getEnd + 1).build())
-        else
-          None
-
-        Seq(rangeBefore, rangeAfter).flatten
-      }
+  def subtractScalarValue(resource: MesosProtos.Resource, amount: Double): Option[MesosProtos.Resource] = {
+    require(resource.getType == MesosProtos.Value.Type.SCALAR)
+    val isMountDiskResource =
+      resource.hasDisk && resource.getDisk.hasSource &&
+        (resource.getDisk.getSource.getType == Source.Type.MOUNT)
+
+    val leftOver: Double = resource.getScalar.getValue - amount
```

Comment:
mesos uses fixed precision for addition/subtraction. would be logical for us to do the same